File: built-in plugins dsp/mb-sampler-woodwind.cpp
/**
 * MB Woodwind Sampler
 * Category : instrument
 * Type     : sampler
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Orchestral woodwinds
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_SAMPLER_WOODWIND_H
#define MB_SAMPLER_WOODWIND_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbSamplerWoodwind : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-sampler-woodwind";
    static constexpr const char* PLUGIN_NAME    = "MB Woodwind Sampler";
    static constexpr const char* PLUGIN_TYPE    = "sampler";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float instrument = 0.5f;  // range [0, 1]
    float breath = 0.5f;  // range [0, 1]
    float volume = 0.75f;  // range [0, 1]
    };

    MbSamplerWoodwind() = default;
    ~MbSamplerWoodwind() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.instrument = std::clamp(params.instrument, 0f, 1f);
        params.breath = std::clamp(params.breath, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Woodwind Sampler
        return input;
    }
};

#endif // MB_SAMPLER_WOODWIND_H

File: built-in plugins dsp/mb-sampler-woodwind.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbSamplerWoodwindPlugin: PluginDefinition = { id: 'mb-sampler-woodwind', slug: 'mb-sampler-woodwind', name: 'MB Woodwind Sampler', category: 'instrument', type: 'sampler', version: '1.0.0', description: 'Orchestral woodwinds', author: 'Max Booster', grade: 'A', oscillators: [], envelope: { attack: 0.08, decay: 0.4, sustain: 0.7, release: 0.4 }, parameters: [{ id: 'instrument', name: 'Instrument', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: false }, { id: 'breath', name: 'Breath', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.75, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { instrument: 0.5, breath: 0.5, volume: 0.75 } };

export default MbSamplerWoodwindPlugin;
